posthog/api/websocket.py

```diff
@@ -0,0 +1,45 @@
+from channels.generic.websocket import AsyncWebsocketConsumer
+from rest_framework.exceptions import Throttled
+from rest_framework.views import APIView
+from rest_framework.request import Request
+from django.contrib.auth.models import AnonymousUser
+from asgiref.sync import sync_to_async
+
+from django.test import RequestFactory
+from typing import cast
+import json
+
+
+class BaseWebsocketConsumer(AsyncWebsocketConsumer):
+    async def check_authentication(self):
+        """Check if the user is authenticated."""
+        user = self.scope.get("user", AnonymousUser())
+
+        if not user.is_authenticated:
+            await self.close(code=401)
+            return False
+        return True
+
+    async def check_throttling(self, data):
+        # Apply throttling
+        throttles = self.get_throttles(data)
+        view = APIView()  # DRF expects a view instance
+        view.request = cast(Request, RequestFactory())
+        view.request.user = self.scope["user"]  # Attach the user object
+
+        try:
+            for throttle in throttles:
+                if not await sync_to_async(throttle.allow_request)(view.request, view):
+                    raise Throttled()
```

Comment:
We could run all of these throttle checks concurrently. For example, something like:
```python
import asyncio

try:
    async with asyncio.TaskGroup():
        for throttle in throttles:
            async def check_throttle():
                if not await sync_to_async(throttle.allow_request)(view.request, view):
                    raise Throttled()
            tg.create_task(check_throttle())
except* Throttled as e:
    ...
```

Reply:
The benefits of this are going to depend on whether the throttle checks are doing some I/O in the background, which seems like that's the case. But like most other async patterns, it will need to be proved by testing.

Reply:
Noted, will do this in a next PR! just want to see whether I can get websockets working at all to start with.